File: CS/XAF.LinqToXpo/DisableStandardActionsLinqListViewController.cs
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.SystemModule;

namespace XAF.LinqToXpo {
    public class DisableStandardActionsForLinqListViewController : ViewController<ListView> {
        private const string LinqCollectionSourceActiveKey = "LinqCollectionSource is used";
        
        private void SetControllerActivity<T>(System.Action<Controller> action) where T : Controller {
            T controller = Frame.GetController<T>();
            if (controller != null) {
                action(controller);
            }
        }

        protected override void OnActivated() {
            base.OnActivated();
            bool flag = !View.Id.EndsWith(LinqCollectionSource.DefaultSuffix);
            SetControllerActivity<ListViewProcessCurrentObjectController>((Controller controller) => { controller.Active[LinqCollectionSourceActiveKey] = flag; });
            SetControllerActivity<DeleteObjectsViewController>((Controller controller) => { controller.Active[LinqCollectionSourceActiveKey] = flag; });
            SetControllerActivity<NewObjectViewController>((Controller controller) => { controller.Active[LinqCollectionSourceActiveKey] = flag; });
            SetControllerActivity<FilterController>((Controller controller) => { controller.Active[LinqCollectionSourceActiveKey] = flag; });
        }
        protected override void OnDeactivated() {
            base.OnDeactivated();
            SetControllerActivity<ListViewProcessCurrentObjectController>((Controller controller) => { controller.Active.RemoveItem(LinqCollectionSourceActiveKey); });
            SetControllerActivity<DeleteObjectsViewController>((Controller controller) => { controller.Active.RemoveItem(LinqCollectionSourceActiveKey); });
            SetControllerActivity<NewObjectViewController>((Controller controller) => { controller.Active.RemoveItem(LinqCollectionSourceActiveKey); });
            SetControllerActivity<FilterController>((Controller controller) => { controller.Active.RemoveItem(LinqCollectionSourceActiveKey); });
        }
    }
}

File: VB/XAF.LinqToXpo/DisableStandardActionsLinqListViewController.vb
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.SystemModule

Namespace XAF.LinqToXpo
    Public Class DisableStandardActionsForLinqListViewController
        Inherits ViewController(Of ListView)

        Private Const LinqCollectionSourceActiveKey As String = "LinqCollectionSource is used"

        Private Sub SetControllerActivity(Of T As Controller)(ByVal action As System.Action(Of Controller))
            Dim controller As T = Frame.GetController(Of T)()
            If controller IsNot Nothing Then
                action(controller)
            End If
        End Sub

        Protected Overrides Sub OnActivated()
            MyBase.OnActivated()
            Dim flag As Boolean = Not View.Id.EndsWith(LinqCollectionSource.DefaultSuffix)
            SetControllerActivity(Of ListViewProcessCurrentObjectController)(Sub(controller As Controller)
                controller.Active(LinqCollectionSourceActiveKey) = flag
            End Sub)
            SetControllerActivity(Of DeleteObjectsViewController)(Sub(controller As Controller)
                controller.Active(LinqCollectionSourceActiveKey) = flag
            End Sub)
            SetControllerActivity(Of NewObjectViewController)(Sub(controller As Controller)
                controller.Active(LinqCollectionSourceActiveKey) = flag
            End Sub)
            SetControllerActivity(Of FilterController)(Sub(controller As Controller)
                controller.Active(LinqCollectionSourceActiveKey) = flag
            End Sub)
        End Sub
        Protected Overrides Sub OnDeactivated()
            MyBase.OnDeactivated()
            SetControllerActivity(Of ListViewProcessCurrentObjectController)(Sub(controller As Controller)
                controller.Active.RemoveItem(LinqCollectionSourceActiveKey)
            End Sub)
            SetControllerActivity(Of DeleteObjectsViewController)(Sub(controller As Controller)
                controller.Active.RemoveItem(LinqCollectionSourceActiveKey)
            End Sub)
            SetControllerActivity(Of NewObjectViewController)(Sub(controller As Controller)
                controller.Active.RemoveItem(LinqCollectionSourceActiveKey)
            End Sub)
            SetControllerActivity(Of FilterController)(Sub(controller As Controller)
                controller.Active.RemoveItem(LinqCollectionSourceActiveKey)
            End Sub)
        End Sub
    End Class
End Namespace
